RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1481 @@
+          rawCharge = (charge - pedestal) * factor + pedestal;
+
+#ifdef HCAL_MAHI_GPUDEBUG
+          printf("first = %d last = %d sipmQ = %f factor = %f rawCharge = %f\n", first, last, sipmq, factor, rawCharge);
+#endif
+        }
+        return rawCharge;
+      }
+
+      // Tried using lambda, but strange error from  with nvcc
+      inline constexpr bool TSenergyCompare(std::pair<unsigned int, float> a, std::pair<unsigned int, float> b) {
+        return a.second > b.second;
+      };
+
+      class kernel_prep1d_sameNumberOfSamples {
+      public:
+        template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+        ALPAKA_FN_ACC void operator()(TAcc const& acc,
+                                      OProductType::View outputGPU,
+                                      IProductTypef01::ConstView f01HEDigis,
+                                      IProductTypef5::ConstView f5HBDigis,
+                                      IProductTypef3::ConstView f3HBDigis,
+                                      HcalMahiConditionsPortableDevice::ConstView mahi,
+                                      HcalSiPMCharacteristicsPortableDevice::ConstView sipmCharacteristics,
+                                      HcalRecoParamWithPulseShapeDevice::ConstView recoParamsWithPS,
+                                      bool const useEffectivePedestals,
+                                      int const sipmQTSShift,
+                                      int const sipmQNTStoSum,
+                                      int const firstSampleShift,
+                                      float const ts4Thresh,
+                                      float* amplitudes,
+                                      float* noiseTerms,
+                                      float* electronicNoiseTerms,
+                                      float* soiSamples,
+                                      int const windowSize) const {
+          auto const nchannelsf015 = f01HEDigis.size() + f5HBDigis.size();
+          auto const startingSample = compute_nsamples<Flavor1>(f01HEDigis.stride()) - windowSize;
+          auto const nchannels = f01HEDigis.size() + f5HBDigis.size() + f3HBDigis.size();
+
+          //first index = groups of channel
+          auto const nchannels_per_block(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[0u]);
+          //2nd index = groups of sample
+          auto const nsamplesForCompute(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[1u]);
+
+          //Loop over all groups of channels
+          for (auto group : uniform_groups_y(acc, nchannels)) {
+            //Loop over each channel, first compute soiSamples for all channels
+            for (auto channel : uniform_group_elements_y(acc, group, nchannels)) {
+              auto const gch = channel.global;
+              if (gch >= nchannels)
+                return;
```

Comment:
Shouldn't the `uniform_group_elements_y()` guarantee `gch < nchannels`?

Reply:
It should.

Reply:
removed the check in https://github.com/cms-sw/cmssw/commit/98367cbd606553c45f30a100a438bedbe880ae55